File: iosApp/Sources/DeviceTestView.swift
import SwiftUI

/// 设备能力自测 — mirrors androidApp DeviceTestView.
/// Direct mode bypasses the JS bridge; full-chain mode goes through
/// harness.js `deviceSelfTest`, exercising the whole bridge path.
struct DeviceTestView: View {

    @Environment(\.dismiss) private var dismiss

    @State private var cases: [DeviceTestCase] = DeviceTestView.buildCases()
    @State private var fullChain = false
    @State private var running = false

    var body: some View {
        VStack(spacing: 0) {
            header
            hint
            ScrollView {
                LazyVStack(spacing: 6) {
                    ForEach(cases.indices, id: \.self) { idx in
                        caseRow(cases[idx])
                            .onTapGesture { runOne(idx) }
                    }
                    unsupportedSection
                }
                .padding(.horizontal, 12)
                .padding(.top, 12)
                .padding(.bottom, 40)
            }
        }
        .background(Theme.background.ignoresSafeArea())
    }

    // ── header ───────────────────────────────────────────────

    private var header: some View {
        VStack(spacing: 0) {
            HStack(spacing: 8) {
                Button {
                    dismiss()
                } label: {
                    Image(systemName: "chevron.left")
                        .font(.system(size: 17, weight: .semibold))
                        .foregroundStyle(Theme.textPrimary)
                        .frame(width: 34, height: 30)
                }
                VStack(alignment: .leading, spacing: 2) {
                    Text("设备能力自测")
                        .font(.system(size: 17, weight: .medium))
                        .foregroundStyle(Theme.textPrimary)
                    Text(fullChain ? "全链路：JavaScriptCore 桥 → native → DeviceBridge" : "直连：DeviceBridge → native API")
                        .font(.system(size: 10))
                        .foregroundStyle(Theme.textHint)
                }
                Spacer(minLength: 8)
                Button {
                    fullChain.toggle()
                } label: {
                    Text(fullChain ? "全链路" : "直连")
                        .font(.system(size: 12, weight: .medium))
                        .foregroundStyle(Theme.primary)
                        .padding(.horizontal, 12)
                        .padding(.vertical, 6)
                        .background(Theme.primaryDim, in: Capsule())
                }
                Button {
                    runAll()
                } label: {
                    Text(running ? "运行中…" : "运行全部")
                        .font(.system(size: 12, weight: .medium))
                        .foregroundStyle(running ? Theme.textHint : Theme.onPrimary)
                        .padding(.horizontal, 12)
                        .padding(.vertical, 6)
                        .background(running ? Theme.surfaceElevated : Theme.primary, in: Capsule())
                }
                .disabled(running)
            }
            .padding(.horizontal, 8)
            .padding(.vertical, 12)
        }
        .background(Theme.surface)
    }

    private var hint: some View {
        Text("逐项调用设备桥：查询类直接校验；交互类（授权/选择器/分享/相机）会拉起系统界面；iOS 相机走系统相机 UI，无静默抓拍。")
            .font(.system(size: 10))
            .foregroundStyle(Theme.textHint)
            .frame(maxWidth: .infinity, alignment: .leading)
            .padding(.horizontal, 16)
            .padding(.vertical, 2)
            .padding(.bottom, 10)
            .background(Theme.surface)
    }

    // ── rows ─────────────────────────────────────────────────

    private func caseRow(_ c: DeviceTestCase) -> some View {
        HStack(alignment: .center, spacing: 8) {
            VStack(alignment: .leading, spacing: 4) {
                HStack(spacing: 8) {
                    Text(c.name)
                        .font(.system(size: 13, weight: .medium))
                        .foregroundStyle(Theme.textPrimary)
                    Text(c.op)
                        .font(.system(size: 10))
                        .foregroundStyle(Theme.textHint)
                }
                if !c.detail.isEmpty {
                    Text(c.detail)
                        .font(.system(size: 10))
                        .foregroundStyle(Theme.textHint)
                        .lineLimit(2)
                        .truncationMode(.tail)
                }
            }
            Spacer(minLength: 8)
            Text(c.state)
                .font(.system(size: 12, weight: .medium))
                .foregroundStyle(Theme.statusColor(c.state))
        }
        .padding(.horizontal, 12)
        .padding(.vertical, 10)
        .background(Theme.surfaceElevated, in: RoundedRectangle(cornerRadius: 10))
    }

    /// Capabilities the iOS bridge does not implement yet — informational only.
    private var unsupportedSection: some View {
        VStack(alignment: .leading, spacing: 6) {
            Text("iOS 端暂未接入")
                .font(.system(size: 11, weight: .medium))
                .foregroundStyle(Theme.textHint)
                .padding(.top, 10)
            ForEach(Self.unsupported, id: \.0) { item in
                HStack(spacing: 8) {
                    VStack(alignment: .leading, spacing: 2) {
                        Text(item.1)
                            .font(.system(size: 13, weight: .medium))
                            .foregroundStyle(Theme.textSecondary)
                        Text(item.0)
                            .font(.system(size: 10))
                            .foregroundStyle(Theme.textHint)
                    }
                    Spacer(minLength: 8)
                    Text("未接入")
                        .font(.system(size: 11))
                        .foregroundStyle(Theme.textHint)
                        .padding(.horizontal, 8)
                        .padding(.vertical, 3)
                        .background(Theme.surfaceElevated, in: Capsule())
                }
                .padding(.horizontal, 12)
                .padding(.vertical, 10)
                .background(Theme.surface.opacity(0.6), in: RoundedRectangle(cornerRadius: 10))
            }
        }
    }

    // ── execution ────────────────────────────────────────────

    private func runOne(_ idx: Int) {
        guard !running else { return }
        running = true
        Task { @MainActor in
            defer { running = false }
            await execute(idx)
        }
    }

    private func runAll() {
        guard !running else { return }
        running = true
        Task { @MainActor in
            defer { running = false }
            for idx in cases.indices { await execute(idx) }
        }
    }

    @MainActor
    private func execute(_ idx: Int) async {
        let engine = LocalEngine.get()
        let op = cases[idx].op
        let argsJson = cases[idx].argsJson
        cases[idx].state = "运行中"
        cases[idx].detail = ""
        do {
            try await engine.ensureStarted()
            let started = Date()
            let obj: [String: Any]
            if fullChain {
                obj = try await engine.deviceFullChainCall(op: op, argsJson: argsJson)
            } else {
                let (_, json) = try await engine.deviceDirectCall(op: op, argsJson: argsJson)
                obj = (try? JSONSerialization.jsonObject(with: Data(json.utf8))) as? [String: Any] ?? [:]
            }
            let took = Int(Date().timeIntervalSince(started) * 1000)
            let errText = obj["error"] as? String ?? ""
            let ok = (obj["ok"] as? Bool ?? true) && errText.isEmpty
            cases[idx].state = ok ? "通过" : "失败"
            cases[idx].detail = "\(took)ms · " + String(Self.describe(obj).prefix(160))
        } catch {
            cases[idx].state = "失败"
            cases[idx].detail = error.localizedDescription
        }
    }

    private static func describe(_ obj: [String: Any]) -> String {
        guard let data = try? JSONSerialization.data(withJSONObject: obj, options: [.sortedKeys]) else {
            return "\(obj)"
        }
        return String(data: data, encoding: .utf8) ?? "\(obj)"
    }

    // ── catalog ──────────────────────────────────────────────

    /// Args follow the *iOS* DeviceBridge contract (op/files keys differ from Android).
    private static func buildCases() -> [DeviceTestCase] {
        var seq = 0
        func c(_ name: String, _ op: String, _ args: String) -> DeviceTestCase {
            seq += 1
            return DeviceTestCase(id: seq, name: name, op: op, argsJson: args)
        }
        return [
            c("状态总览", "status", "{}"),
            c("权限查询 — 七域状态", "permissions", #"{"op":"list"}"#),
            c("权限申请 — 相机授权弹窗", "permissions", #"{"op":"request","name":"camera"}"#),
            c("剪贴板写入", "clipboard", #"{"op":"write","text":"harnest-device-test"}"#),
            c("剪贴板读取 — 校验回读", "clipboard", #"{"op":"read"}"#),
            c("沙箱写文件", "files", #"{"op":"writeFile","path":"device-test/hello.txt","data":"hello from device bridge"}"#),
            c("沙箱读文件 — 校验回读", "files", #"{"op":"readFile","path":"device-test/hello.txt"}"#),
            c("沙箱列目录", "files", #"{"op":"readdir","path":"device-test"}"#),
            c("文件选择器 — 系统文档选择", "files", #"{"op":"pick"}"#),
            c("图片选择器 — 系统照片选择", "photos", #"{"op":"pick"}"#),
            c("相机 — 系统相机 UI", "camera", #"{"op":"capture"}"#),
            c("网络 — 连通性概览", "network", "{}"),
            c("设备信息", "deviceinfo", "{}"),
            c("触感震动", "vibrate", #"{"style":"heavy"}"#),
            c("分享面板 — 系统分享", "share", #"{"text":"harnest device test"}"#),
        ]
    }

    private static let unsupported: [(String, String)] = [
        ("contacts", "通讯录 — 读写联系人"),
        ("calendar", "日历 — 事件读写"),
        ("call / sms", "拨号 / 短信 — 预填拉起"),
        ("mail", "邮件 — 拉起撰写"),
        ("recorder", "录音 — 开始 / 停止"),
        ("app", "应用 — 列表 / 跳转"),
        ("location", "定位 — 一次定位"),
        ("scheduler", "提醒 / 系统设置 / GUI / 调度"),
    ]
}

struct DeviceTestCase: Identifiable {
    let id: Int
    let name: String
    let op: String
    let argsJson: String
    var state: String = "待测"
    var detail: String = ""
}

File: androidApp/app/src/main/java/com/harnest/app/ui/DeviceTestView.kt
package com.harnest.app.ui

import android.widget.Toast
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.systemBarsPadding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.harnest.app.service.LocalEngine
import kotlinx.coroutines.launch
import org.json.JSONObject

class DeviceCase(val name: String, val op: String, val argsJson: String) {
    var state: String = "待测"
    var detail: String = ""
}

@Composable
fun DeviceTestView(onClose: () -> Unit) {
    val c = harnessColors()
    val context = LocalContext.current
    val scope = rememberCoroutineScope()
    val cases = remember { buildCases() }
    var rev by remember { mutableStateOf(0) }
    var fullChain by remember { mutableStateOf(false) }
    var lastId by remember { mutableStateOf<String?>(null) }
    var running by remember { mutableStateOf(false) }

    fun stateColor(state: String) = when (state) {
        "通过" -> c.success
        "失败" -> c.error
        "运行中" -> c.warning
        else -> c.textHint
    }

    suspend fun execute(case: DeviceCase) {
        case.state = "运行中"
        case.detail = ""
        rev++
        try {
            LocalEngine.get().ensureStarted(context)
            val args = if (case.argsJson.contains("__LAST__")) {
                val last = lastId
                if (last.isNullOrEmpty()) {
                    throw IllegalArgumentException("缺少前置新建结果（先执行对应的新建用例）")
                }
                case.argsJson.replace("__LAST__", last)
            } else {
                case.argsJson
            }
            val started = System.currentTimeMillis()
            val result = if (fullChain) {
                LocalEngine.get().deviceFullChainCall(case.op, args)
            } else {
                val (_, json) = LocalEngine.get().deviceDirectCall(case.op, args)
                JSONObject(json)
            }
            val took = System.currentTimeMillis() - started
            val idVal: Any? = result.opt("id")
            val idStr = when (idVal) {
                is String -> idVal
                is Int -> idVal.toString()
                is Long -> idVal.toString()
                else -> null
            }
            if (!idStr.isNullOrEmpty()) lastId = idStr
            val ok = result.optBoolean("ok", true) && result.optString("error").isBlank()
            case.state = if (ok) "通过" else "失败"
            case.detail = if (case.op == "recorder" && args.contains("\"start\"")) {
                "${took}ms · 录音中…（稍后执行「停止」返回文件）"
            } else {
                "${took}ms · " + result.toString().take(160)
            }
        } catch (e: Throwable) {
            case.state = "失败"
            case.detail = e.message ?: e.javaClass.simpleName
        }
        rev++
    }

    fun runOne(case: DeviceCase) {
        if (running) {
            Toast.makeText(context, "有用例正在运行", Toast.LENGTH_SHORT).show()
            return
        }
        running = true
        scope.launch {
            try {
                execute(case)
            } finally {
                running = false
            }
        }
    }

    fun runAll() {
        if (running) {
            Toast.makeText(context, "有用例正在运行", Toast.LENGTH_SHORT).show()
            return
        }
        running = true
        scope.launch {
            try {
                for (case in cases) execute(case)
            } finally {
                running = false
            }
        }
    }

    Column(
        Modifier
            .fillMaxSize()
            .background(c.background)
            .systemBarsPadding()
    ) {
        Row(
            Modifier
                .fillMaxWidth()
                .background(c.surface)
                .padding(horizontal = 8.dp, vertical = 12.dp),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            Text(
                "‹",
                color = c.textPrimary,
                fontSize = 22.sp,
                modifier = Modifier
                    .clickable(onClick = onClose)
                    .padding(horizontal = 10.dp, vertical = 2.dp),
            )
            Column(Modifier.weight(1f)) {
                Text(
                    "设备能力自测",
                    color = c.textPrimary,
                    fontSize = 17.sp,
                    fontWeight = FontWeight.Medium,
                )
                Text(
                    if (fullChain) "全链路：QuickJS 桥 → native → DeviceBridge" else "直连：DeviceBridge → native API",
                    color = c.textHint,
                    fontSize = 10.sp,
                )
            }
            Text(
                if (fullChain) "全链路" else "直连",
                color = c.primary,
                fontSize = 12.sp,
                modifier = Modifier
                    .background(c.primary.copy(alpha = 0.12f), RoundedCornerShape(12.dp))
                    .clickable { fullChain = !fullChain }
                    .padding(horizontal = 12.dp, vertical = 6.dp),
            )
            Spacer(Modifier.width(8.dp))
            Text(
                if (running) "运行中…" else "运行全部",
                color = if (running) c.textHint else c.onPrimary,
                fontSize = 12.sp,
                modifier = Modifier
                    .background(if (running) c.surfaceElevated else c.primary, RoundedCornerShape(12.dp))
                    .clickable { runAll() }
                    .padding(horizontal = 12.dp, vertical = 6.dp),
            )
        }
        Text(
            "逐项调用设备桥：查询类直接校验；交互类（授权/选择器/拨号/相机等）会拉起系统界面；新建 + 删除成对出现。",
            color = c.textHint,
            fontSize = 10.sp,
            modifier = Modifier
                .fillMaxWidth()
                .background(c.surface)
                .padding(horizontal = 16.dp, vertical = 2.dp)
                .padding(bottom = 10.dp),
        )
        LazyColumn(
            Modifier
                .weight(1f)
                .fillMaxWidth(),
            contentPadding = androidx.compose.foundation.layout.PaddingValues(
                start = 12.dp, end = 12.dp, top = 12.dp, bottom = 40.dp,
            ),
            verticalArrangement = Arrangement.spacedBy(6.dp),
        ) {
            items(cases.size, key = { it }) { idx ->
                rev
                val case = cases[idx]
                Row(
                    Modifier
                        .fillMaxWidth()
                        .background(c.surfaceElevated, RoundedCornerShape(10.dp))
                        .clickable(enabled = !running) { runOne(case) }
                        .padding(horizontal = 12.dp, vertical = 10.dp),
                    verticalAlignment = Alignment.CenterVertically,
                ) {
                    Column(Modifier.weight(1f)) {
                        Row(verticalAlignment = Alignment.CenterVertically) {
                            Text(
                                case.name,
                                color = c.textPrimary,
                                fontSize = 13.sp,
                                fontWeight = FontWeight.Medium,
                            )
                            Spacer(Modifier.width(8.dp))
                            Text(
                                "${case.op}",
                                color = c.textHint,
                                fontSize = 10.sp,
                            )
                        }
                        if (case.detail.isNotEmpty()) {
                            Text(
                                case.detail,
                                color = c.textHint,
                                fontSize = 10.sp,
                                maxLines = 2,
                                overflow = TextOverflow.Ellipsis,
                            )
                        }
                    }
                    Spacer(Modifier.width(8.dp))
                    Text(
                        case.state,
                        color = stateColor(case.state),
                        fontSize = 12.sp,
                        fontWeight = FontWeight.Medium,
                    )
                }
            }
        }
    }
}

private fun buildCases(): List<DeviceCase> = listOf(
    DeviceCase("状态总览", "status", "{}"),
    DeviceCase("权限查询 — 四域状态", "permissions", """{"action":"query"}"""),
    DeviceCase("权限申请 — 弹系统授权框", "permissions", """{"action":"request"}"""),
    DeviceCase("剪贴板写入", "clipboard", """{"op":"write","text":"harnest-device-test"}"""),
    DeviceCase("剪贴板读取 — 校验回读", "clipboard", """{"op":"read"}"""),
    DeviceCase("沙箱写文件", "files", """{"op":"sandboxWrite","path":"device-test/hello.txt","text":"hello from device bridge"}"""),
    DeviceCase("沙箱读文件 — 校验回读", "files", """{"op":"sandboxRead","path":"device-test/hello.txt"}"""),
    DeviceCase("沙箱列目录", "files", """{"op":"sandboxList","path":"device-test"}"""),
    DeviceCase("文件选择器 — 系统文档选择", "files", """{"op":"pick"}"""),
    DeviceCase("图片选择器 — 系统照片选择", "photos", """{"op":"pick"}"""),
    DeviceCase("通讯录查询 — 前 20 条", "contacts", """{"op":"query","limit":20}"""),
    DeviceCase("通讯录新建 — 测试联系人", "contacts", """{"op":"create","name":"Harnest 测试","phones":["13800000000"]}"""),
    DeviceCase("通讯录删除 — 按查询结果", "contacts", """{"op":"delete","id":"__LAST__"}"""),
    DeviceCase("日历查询 — 前后 7 天", "calendar", """{"op":"query"}"""),
    DeviceCase("日历新建 — 测试事件", "calendar", """{"op":"create","title":"Harnest 测试事件"}"""),
    DeviceCase("日历删除 — 按新建结果", "calendar", """{"op":"delete","id":"__LAST__"}"""),
    DeviceCase("拨号盘 — 预填号码", "call", """{"op":"dial","number":"13800138000"}"""),
    DeviceCase("短信 — 预填收件人与内容", "call", """{"op":"sms","number":"13800138000","body":"harnest test"}"""),
    DeviceCase("短信读取 — 受限降级说明", "sms", """{"op":"query","days":30}"""),
    DeviceCase("邮件 — 拉起撰写界面", "mail", """{"to":"test@example.com","subject":"Harnest 测试"}"""),
    DeviceCase("相机 — 自动拍照（无 UI）", "camera", """{"op":"capture"}"""),
    DeviceCase("录音开始 — 5 秒上限", "recorder", """{"op":"start","maxSeconds":5}"""),
    DeviceCase("录音停止 — 返回文件", "recorder", """{"op":"stop"}"""),
    DeviceCase("应用列表 — 可启动应用", "app", """{"op":"list"}"""),
    DeviceCase("打开系统设置", "app", """{"op":"open","package":"com.android.settings"}"""),
)
